service/matching/matching_engine_test.go

```diff
@@ -2322,6 +2322,124 @@ func (s *matchingEngineSuite) TestAddActivityTask_ForVersionedWorkflows_Silently
 	s.Require().NoError(err)
 }
 
+func (s *matchingEngineSuite) TestPersistUnknownBuildId_Poll() {
+	namespaceId := namespace.ID(uuid.New())
+	tl := "makeToast"
+	tlID := newTestTaskQueueID(namespaceId, tl, enumspb.TASK_QUEUE_TYPE_WORKFLOW)
+
+	scope := tally.NewTestScope("test", nil)
+	s.matchingEngine.metricsHandler = metrics.NewTallyMetricsHandler(metrics.ClientConfig{}, scope)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
+	defer cancel()
+
+	_, err := s.matchingEngine.getTask(ctx, tlID, normalStickyInfo, &pollMetadata{
+		workerVersionCapabilities: &commonpb.WorkerVersionCapabilities{
+			BuildId:       "unknown",
+			UseVersioning: true,
+		},
+	})
+	s.Error(err) // deadline exceeded or canceled
+
+	unknownCtr := scope.Snapshot().Counters()["test.unknown_build_polls+namespace="+matchingTestNamespace+",operation=TaskQueueMgr,task_type=Workflow,taskqueue=makeToast"]
```

Comment:
Not for this PR, but I think we should have a better API for querying this. It's not great for it to be both:
1. Sensitive to the ordering of the tags
2. Require you to specify all tags--even ones you may not necessarily care about